### ISSUE TITLE
Some wrong comes when I use the SCTransformed Data from Seruat

### ISSUE THREAD

**Issue by Midsummer723** [LTLA/SingleR#98]
**Question1**
I used the SCTransform Function in Seurat 3. Then I got the "integrated data"
When I run the 
example_sce <- logNormCounts(example_sce)
Error in assay(x, exprs_values) : 
  'assay(<SingleCellExperiment>, i="character", ...)' invalid subscript 'i'
'counts' not in names(assays(<SingleCellExperiment>))

**Question2:**
I use a very big data.
 I got a result of label.fine
when I run the 
plotScoreHeatmap(pred.hpca)
Error in hclust(d, method = method) : size cannot be NA nor exceed 65536
The plotScoreHeatmap function seems to need a lot of memory.
At my case I use like 148GB.

**Comment by Midsummer723** [comment]
Actually I tried another data, which could work fine in the ref ImmGenData at $label.mian.
But I try run it again, this time it does not work anymore. Is the function “ $label.mian” cannot work ?

**Comment by LTLA** [comment]
Before I answer your questions, I will note that it is usually good etiquette to:

- Have a minimal reproducible example.
- Show your session information.
- Format your post properly with markdown.
- Avoid obvious spelling mistakes.

![stuff](https://user-images.githubusercontent.com/8166669/75518212-3e748e00-59b5-11ea-84b8-884b4975f961.png)

An occasional error is tolerable. But if you have mistakes all over the place, it appears sloppy. And that leads me to conclude that your code is also sloppy. And that causes me to write off your request. (And yes, grant funding agencies and HR interviewer screens work the same way.)

> **Questin1**
> I used the SCTransform Function in Seruat 3. Then I got the "integrated data"
> When I run the
> example_sce <- logNormCounts(example_sce)
> Error in assay(x, exprs_values) :
> 'assay(, i="character", ...)' invalid subscript 'i'
> 'counts' not in names(assays())

Not sure there's much more to say here. As it says, you don't have a `counts` field in your `assays()`.

I don't know why you're using the **sctransform**'ed data, just give it the original counts.

> **Question2:**
> I use a very big data.
> I got a result of label.fine
> when I run the
> plotScoreHeatmap(pred.hpca)
> Error in hclust(d, method = method) : size cannot be NA nor exceed 65536
> The plotScoreHeatmap functino seems to need a lot of memory.
> At my case I use like 148GB.

Set `cells.order=order(pred.hpca$label)` to ensure that `pheatmap` does not attempt clustering.

@dtm2451 Whatever the default is, it should probably not attempt clustering of the cells.

**Comment by dtm2451** [comment]
Noted.  I'll turn clustering off by default for large datasets.

For Question 1, @Midsummer723 it's not clear why you are trying to use logNormCounts().  For one, you can provide SingleR raw counts and it should work just fine. I believe `GetAssayData(seurat, slot = "raw.data")` gets the raw counts.  But more than that, SCTransform data is basically log normalized already, plus more.  To use that data directly, you can provide `GetAssayData(seruat, assay = "SCT", slot = "data")`

**Comment by Midsummer723** [comment]
I appreciate that  you can point out my fault. I will pay more attention to it,
I use the SCTransform integration. Because I have 30 patients' samples and I want to compare them which can show the differences. Also it can lead to the data is too big. 
So actually I want to know if there is a better way to display the data with heatmap or other ways.
Thanks a lot.

**Comment by LTLA** [comment]
@dtm2451 My impression is that we should just have an explicit `cluster_cols=FALSE` argument that people have to actively turn modify. Otherwise, the best default is probably to order cells by their labels, if cluster information has not been provided.

@Midsummer723 I do not know how **sctransform**'ed data will behave with `SingleR()`. AFAIK it is not a monotonic transformation within each cell, which makes it hard to say that `SingleR()` will work in the same manner as it does for (log-transformed) counts.

I don't see a good reason for you to use **sctransform** here. Just use the raw counts.

**Comment by dtm2451** [comment]
@Midsummer723 I think it is great that you are a medical student analyzing single-cell data!

@LTLA I've made an explicit `cluster_cols=FALSE` argument and set the default ordering to be by labels with #100

My understanding of **sctransform** data is that its final structure resembles log-transformed counts.  I haven't performed extensive testing, but I have found the "SCT" assay to serve well in `SingleR(test)` with either `ref = BlueprintEncodeData()` or with `ref = SCT_data_of_manually_annotated_seurat`.

**Comment by LTLA** [comment]
I will provide a simple demonstration of **sctransform**'s unpredictability. Let's create a very simple mock dataset where we know exactly what's going on. Firstly, we define some gene abundances: 

```r
set.seed(1000)
ngenes <- 10000
mu <- 2^rnorm(10000)
```

Now we create the population means. In this case, I am considering a very straightforward case of two populations that only differ in the first 100 genes that exhibit a 10-fold difference. As a control I will throw in another 100 genes that do not exhibit any DE between populations.

```r
pop.means <- matrix(mu, nrow=ngenes, ncol=2)
pop.means[1:100,] <- matrix(c(0.5, 5), 100, ncol=2, byrow=TRUE)
pop.means[100+1:100,] <- matrix(c(0.5, 0.5), 100, ncol=2, byrow=TRUE)
```

Right, now let's create some cells. We'll just spawn 500 cells from each population, nothing too hard there. To keep things somewhat interesting, we will add a library size difference between our two populations (otherwise the simulation would be boring and there wouldn't even be any point running **sctransform**). In this case, we introduce a mild 2-fold difference in library size, not particularly extreme given the range of total RNA content between cell types.

```r
g <- rep(1:2, each=500)
cell.means <- pop.means[,g,drop=FALSE]
cell.means[,g==1] <- cell.means[,g==1]/2
```

Finally, we simulate the count data. This is almost as simple as can be, using a negative binomial distribution without any mean-dispersion trend. We do need to give it some dimnames, though, otherwise **sctransform** will complain.

```r
y <- matrix(rnbinom(length(cell.means), mu=cell.means, size=100), 
    nrow=ngenes, ncol=length(g),
    dimnames=list(1:nrow(cell.means), 1:ncol(cell.means)))
```

Okey dokey. Let's hit up **sctransform** for some corrected UMIs! While we know the originating population in this simulation, we wouldn't obviously know it in our actual analyses (otherwise why bother doing the analysis?), so I'm not going to put it in as a latent factor.

```r
library(sctransform)
out <- vst(y, return_corrected_umi=TRUE)
```

Now let's have a look at our two sets of genes. For brevity, I'll take the mean corrected UMI for each cell, and then compute the mean across cells within each of our populations.

```r
lapply(split(Matrix::colMeans(out$umi_corrected[1:100,]), g), mean)
## $`1`
## [1] 0.35648
## 
## $`2`
## [1] 3.88156

lapply(split(Matrix::colMeans(out$umi_corrected[1:100+100,]), g), mean)
## $`1`
## [1] 0.25114
## 
## $`2`
## [1] 0.48292
```

So now a bit of magic has happened. As I predicted, the transformation applied by **sctransform** is not monotonic, even on average; we can see that the first 100 genes in the first population now have systematically higher average expression values compared to the second set of 100 genes, despite the fact that they both started off with a mean count of 0.5! This is potentially troublesome for **SingleR** and other Spearman-based methods as the original ranking is not respected.

Even more magical is the introduction of a ~2-fold difference in the corrected UMI between populations for the second set of 100 genes. Recall that these genes should not have been DE between populations, so it's a bit of a mystery where this difference comes from. And before you say that it's caused by the 2-fold difference in the library sizes, I'll remind you that we're working on the _corrected_ UMIs where this difference should have already been removed. If I have to do another round of library size normalization... then what's the point of running **sctransform**? Obviously, seeing this mysterious DE does not inspire confidence in the detection of marker genes.

Now, maybe the biological signal is so strong that none of this makes a difference. But if that's the case, why not run the analysis on the raw counts? You don't have to worry about what **sctransform** (or any transformation) is doing to your data; you don't have to worry about results changing when the **sctransform** maintainers decide to update their algorithm; and you don't have to sit around waiting for **sctransform** to run. Importantly, by using the raw counts, you preserve a useful property of classification algorithms - that each cell is classified independently of each other cell. This is nice from a logistical perspective (doesn't matter whether you classify first and filter later) and a statistical perspective (you can actually do analyses with independent observations).

**Comment by Midsummer723** [comment]
Thanks for your patience! It was very inspirational to me. I will try this when doing my analysis.
I am appreciate your help.

**Comment by dtm2451** [comment]
@LTLA thanks very much for the detailed explanation!

**Comment by LTLA** [comment]
Closed by #100.
